azext_iot/tests/iothub/message_endpoint/conftest.py

```diff
@@ -0,0 +1,464 @@
+        desired_sys_identity = hub_marker.kwargs.get("sys_identity")
+        desired_user_identity = hub_marker.kwargs.get("user_identity")
+
+    if desired_sys_identity:
+        base_create_command += " --mi-system-assigned"
+    if desired_user_identity and provisioned_user_identity:
+        user_identity_id = provisioned_user_identity["id"]
+        base_create_command += f" --mi-user-assigned {user_identity_id}"
+    if desired_tags:
+        base_create_command = base_create_command + f" --tags {desired_tags}"
+    if desired_location:
+        base_create_command = base_create_command + f" -l {desired_location}"
+
+    return cli.invoke(base_create_command).as_json()
+
+
+def _iot_hub_removal(name):
+    delete_result = cli.invoke(f"iot hub delete -n {name} -g {RG}")
+    if not delete_result.success():
+        logger.error(f"Failed to delete iot hub resource {name}.")
+
+
+@pytest.fixture(scope="module")
+def provisioned_user_identity_module() -> Optional[dict]:
+    result = _user_identity_provisioner()
+    yield result
+    if result:
+        _user_identity_removal(result["name"])
+
+
+def _user_identity_provisioner():
+    name = generate_hub_depenency_id()
+    return cli.invoke(
+        f"identity create -n {name} -g {RG}"
+    ).as_json()
+
+
+def _get_role_assignments(scope, role):
+    return cli.invoke(
+        f'role assignment list --scope "{scope}" --role "{role}"'
+    ).as_json()
+
+
+def _user_identity_removal(name):
+    delete_result = cli.invoke(f"identity delete -n {name} -g {RG}")
+    if not delete_result.success():
+        logger.error(f"Failed to delete user identity resource {name}.")
+
+
+def _assign_rbac_role(assignee: str, scope: str, role: str, max_tries: int = 10):
+    tries = 0
+    while tries < max_tries:
+        role_assignments = _get_role_assignments(scope, role)
+        role_assignment_principal_ids = [assignment["principalId"] for assignment in role_assignments]
+        if assignee in role_assignment_principal_ids:
+            break
+        # else assign role to scope and check again
+        cli.invoke(
+            'role assignment create --assignee "{}" --role "{}" --scope "{}"'.format(
+                assignee, role, scope
+            )
+        )
+        sleep(10)
+        tries += 1
+
+
+@pytest.fixture(scope="module")
+def provisioned_storage_with_identity_module(request, provisioned_iot_hub_module, provisioned_storage_module):
+    role = "Storage Blob Data Contributor"
+    scope = provisioned_storage_module["storage"]["id"]
+    hub_principal_id = provisioned_iot_hub_module["identity"]["principalId"]
+    user_identities = list(provisioned_iot_hub_module["identity"]["userAssignedIdentities"].values())
+    user_id = user_identities[0]["principalId"]
+    _assign_rbac_role(assignee=hub_principal_id, scope=scope, role=role)
+    _assign_rbac_role(assignee=user_id, scope=scope, role=role)
+    yield provisioned_iot_hub_module, provisioned_storage_module
```

Comment:
should we do tuples or dicts? dicts are easier and nicer but tuples are smaller

Reply:
If it's only returning two values in a test, I think a tuple is fine for now. It helps us not have to remember named properties, and we can just index the properties (again, since there's only two, `[0]` and `[1]` are fine)